lib/Transforms/IPO/StructAnalysisCloseProximity.cpp

```diff
@@ -0,0 +1,1004 @@
+      errs() << "Update CPG between " << Src << " and " << Dest
+             << " with count " << DEBUG_PRINT_COUNT(C) << " and distance "
+             << DEBUG_PRINT_DIST(D) << "\n";
+    }
+    DEBUG_WITH_TYPE(
+        DEBUG_TYPE_CPG,
+        dbgs() << " After update: "
+               << "(" << DEBUG_PRINT_COUNT(CPT[Src - 1][Dest - 1].first) << ","
+               << DEBUG_PRINT_DIST(CPT[Src - 1][Dest - 1].second) << ")\n");
+  }
+}
+
+// Wrapper function to update a cell in CloseProximityTable
+void CloseProximityBuilder::updateCPG(FieldNumType Src, FieldNumType Dest,
+                                      ExecutionCountType C, DataBytesType D) {
+  DEBUG_WITH_TYPE(DEBUG_TYPE_CPG,
+                  dbgs() << "Will update CPG between " << Src << " and " << Dest
+                         << " with count " << DEBUG_PRINT_COUNT(C)
+                         << " and distance " << DEBUG_PRINT_DIST(D) << ":\n");
+  updateCPT(Src, Dest, C, D, CloseProximityTable);
+}
+
+// Wrapper function to update a cell in GoldCPT
+void CloseProximityBuilder::updateGoldCPG(FieldNumType Src, FieldNumType Dest,
+                                          ExecutionCountType C,
+                                          DataBytesType D) {
+  DEBUG_WITH_TYPE(DEBUG_TYPE_CPG,
+                  dbgs() << "Will update Gold CPG between " << Src << " and "
+                         << Dest << " with count " << DEBUG_PRINT_COUNT(C)
+                         << " and distance " << DEBUG_PRINT_DIST(D) << ":\n");
+  updateCPT(Src, Dest, C, D, GoldCPT);
+}
+
+// Core code to update CP relations along a path from Node From in its subtree
+// Node To is a node on its subtree and the path ends at either a leaf node or
+// a back edge to a node on the path. Every pair of (From, To) on the path will
+// be updated. Edge Arc is used to keep record of the current Edge on the path
+// to be calculated.
+// There are possible two ways to calculate:
+// 1. If the next Edge on the path is a collapsed Edge, the path ends here and
+//    update CP relations between Node From and all collapsed entries on the
+//    Edge. When updating, make sure to take only a portion of the count on the
+//    collapsed entry when the collapsed Edge is one of the incoming edge of the
+//    node.
+// 2. If the next Edge is not collapsed, update CP relation between From and To
+//    and move on to the destination node of the Edge. When moving on,
+//    considering the current node might have multiple exits, take a portion of
+//    the weight is important to ensure correctness. Call
+//    updatePairByConnecting() to only take a portion count when recursively
+//    calling the function.
+// The following example shows overall guidance to guarantee correctness
+// when updating counts: When updating CP relations between A and D,
+//
+//       Count1      Count5     Count8
+//    A --------- B -------- C -------- D
+// Suppose A has two outgoing edges: Count1 and Count2, B has two incoming
```

Comment:
Will update comment to elaborate it more. The comment is used to illustrate how to update count because distance is obvious when calculating along a path. That's why I directly put count on the edge...